lib/value/src/kind.rs

```diff
@@ -0,0 +1,95 @@
+//! The `kind` module has all relevant types related to progressive type checking.
+
+mod accessor;
+mod builder;
+mod collection;
+mod comparison;
+mod conversion;
+pub mod insert;
+pub mod merge;
+
+pub use builder::EmptyKindError;
+pub use collection::{Collection, Field, Index};
+
+/// The type (kind) of a given value.
+///
+/// This struct tracks the known states a type can have. By allowing one type to have multiple
+/// states, the type definition can be progressively refined.
+///
+/// At the start, a type is in the "any" state, meaning its type can be any of the valid states, as
+/// more information becomes available, states can be removed, until one state is left.
+///
+/// A state without any type information (e.g. all fields are `None`) indicates no type information
+/// can be inferred from the value. This is usually a programming error, but it's a valid state for
+/// this library to expose.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Kind {
+    bytes: Option<()>,
+    integer: Option<()>,
+    float: Option<()>,
+    boolean: Option<()>,
+    timestamp: Option<()>,
+    regex: Option<()>,
+    null: Option<()>,
```

Comment:
Just a note that I started with booleans, but preferred having all fields be `Option`s, since the collection fields needed that anyway. These fields aren't public, but it made for a more consistent internal API.

Also, `size_of::<bool>() == size_of::<Option<()>>()`, so there's no performance loss (not that it matters too much for this type and the way it is used).